File: packages/server/src/domains/environments/service.ts
import type { Environment, CreateEnvironmentInput, UpdateEnvironmentInput } from '@flick/shared';
import * as repo from './repo.js';
import { NotFoundError, ConflictError } from '../../shared/errors.js';

export async function listEnvironments(projectId: string): Promise<Environment[]> {
  return repo.findByProject(projectId);
}

export async function getEnvironment(id: string): Promise<Environment> {
  const env = await repo.findById(id);
  if (!env) throw new NotFoundError('Environment', id);
  return env;
}

export async function createEnvironment(
  projectId: string,
  input: CreateEnvironmentInput,
): Promise<Environment> {
  const existing = await repo.findBySlug(projectId, input.slug);
  if (existing) throw new ConflictError(`Environment with slug '${input.slug}' already exists`);
  return repo.create(projectId, input);
}

export async function updateEnvironment(id: string, input: UpdateEnvironmentInput): Promise<Environment> {
  const env = await repo.update(id, input);
  if (!env) throw new NotFoundError('Environment', id);
  return env;
}

export async function deleteEnvironment(id: string): Promise<void> {
  const deleted = await repo.remove(id);
  if (!deleted) throw new NotFoundError('Environment', id);
}

File: packages/server-rust/src/domains/environments/service.rs
use sqlx::PgPool;
use uuid::Uuid;

use crate::error::AppError;

use super::repo::{self, EnvironmentRow};

pub struct CreateEnvironmentInput {
    pub name: String,
    pub slug: String,
    pub color: String,
    pub sort_order: i32,
}

pub struct UpdateEnvironmentInput {
    pub name: Option<String>,
    pub color: Option<String>,
    pub sort_order: Option<i32>,
}

pub async fn list_environments(
    db: &PgPool,
    project_id: Uuid,
) -> Result<Vec<EnvironmentRow>, AppError> {
    repo::find_by_project(db, project_id)
        .await
        .map_err(AppError::from)
}

pub async fn get_environment(db: &PgPool, id: Uuid) -> Result<EnvironmentRow, AppError> {
    repo::find_by_id(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Environment", Some(&id.to_string())))
}

pub async fn create_environment(
    db: &PgPool,
    project_id: Uuid,
    input: CreateEnvironmentInput,
) -> Result<EnvironmentRow, AppError> {
    // Check for slug conflict within the project
    if repo::find_by_slug(db, project_id, &input.slug)
        .await?
        .is_some()
    {
        return Err(AppError::Conflict(format!(
            "Environment with slug '{}' already exists in this project",
            input.slug
        )));
    }

    repo::create(
        db,
        project_id,
        &input.name,
        &input.slug,
        &input.color,
        input.sort_order,
    )
    .await
    .map_err(AppError::from)
}

pub async fn update_environment(
    db: &PgPool,
    id: Uuid,
    input: UpdateEnvironmentInput,
) -> Result<EnvironmentRow, AppError> {
    repo::update(
        db,
        id,
        input.name.as_deref(),
        input.color.as_deref(),
        input.sort_order,
    )
    .await?
    .ok_or_else(|| AppError::not_found("Environment", Some(&id.to_string())))
}

pub async fn delete_environment(db: &PgPool, id: Uuid) -> Result<Uuid, AppError> {
    repo::remove(db, id)
        .await?
        .ok_or_else(|| AppError::not_found("Environment", Some(&id.to_string())))
}
